lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -4634,6 +4634,203 @@ class DecomposeAtenTypeAsOp : public OpRewritePattern<AtenTypeAsOp> {
+          loc, Torch::ListType::get(Torch::IntType::get(context)), dimValues);
+      newInput = rewriter.create<Torch::AtenPermuteOp>(
+          loc,
+          inputType.getWithSizesAndDtype(permutedSizes,
+                                         inputType.getOptionalDtype()),
+          input, dimValueList);
+    } else {
+      newInput = input;
+      for (int i = 0; i < inputRank; i++) {
+        dims.emplace_back(i);
+      }
+    }
+
+    // manually generate new indices.
+    SmallVector<Value> listElements(inputRank);
+    Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+    auto int64Dtype = getDtypeIntValueForType(
+        rewriter, loc, rewriter.getIntegerType(64, true));
+
+    int64_t trailingDimCnt = 0;
+    int64_t i;
+    // handle trailing none index.
+    for (i = inputRank - 1; i >= 0; --i) {
+      int64_t oldI = dims[i];
+      if (!indexUsed[oldI]) {
+        auto resultType = ValueTensorType::get(
+            context, {inputSizes[oldI]}, rewriter.getIntegerType(64, true));
+        auto dim = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(i));
+        auto end = rewriter.create<Torch::AtenSizeIntOp>(loc, newInput, dim);
+        auto v = rewriter.create<Torch::AtenArangeOp>(
+            loc, resultType, end, int64Dtype, none, none, none);
+        auto vInfo =
+            unsqueezeTensorAtTrailingDim(op, rewriter, v, trailingDimCnt);
+        if (failed(vInfo)) {
+          return rewriter.notifyMatchFailure(op, "failed to unsqueeze tensor");
+        }
+        listElements[i] = *vInfo;
+        trailingDimCnt++;
+      } else
+        break;
+    }
+    // handle non-none index in between.
+    for (; i >= 0; --i) {
+      int64_t oldI = dims[i];
+      if (indexUsed[oldI]) {
+        auto vInfo = unsqueezeTensorAtTrailingDim(op, rewriter, indices[oldI],
+                                                  trailingDimCnt);
+        if (failed(vInfo)) {
+          return rewriter.notifyMatchFailure(op, "failed to unsqueeze tensor");
+        }
+        listElements[i] = *vInfo;
+      } else
+        break;
+    }
+
+    // handle possible leading none dimensions.
+    for (; i >= 0; --i) {
+      int64_t oldI = dims[i];
+      if (!indexUsed[oldI]) {
+        auto resultType = ValueTensorType::get(
+            context, {inputSizes[oldI]}, rewriter.getIntegerType(64, true));
+        auto dim = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(i));
+        auto end = rewriter.create<Torch::AtenSizeIntOp>(loc, newInput, dim);
+        auto v = rewriter.create<Torch::AtenArangeOp>(
+            loc, resultType, end, int64Dtype, none, none, none);
+        auto vInfo =
+            unsqueezeTensorAtTrailingDim(op, rewriter, v, outputRank - 1 - i);
+        if (failed(vInfo)) {
+          return rewriter.notifyMatchFailure(op, "failed to unsqueeze tensor");
+        }
+        listElements[i] = *vInfo;
```

Comment:
this code is very similar to the handling of trailing nones above. It should be refactored